rad/src/operators/array.rs

```diff
@@ -111,23 +115,40 @@ pub fn map(input: &RadonArray, args: &[Value]) -> Result<RadonTypes, RadError> {
     };
     let subscript = unpack_subscript(&args[0]).map_err(subscript_err)?;
 
-    let mut result = vec![];
+    let mut reports = vec![];
+    let mut results = vec![];
+
+    let settings = RadonScriptExecutionSettings::enable_all();
     for item in input.value() {
-        // FIXME: add support for bubbling up errors thrown in subcontexts.
-        // FIXME: use the new `execute_radon_script` instead.
-        result.push(execute_contextfree_radon_script(
-            item,
-            subscript.as_slice(),
-        )?);
+        let report = execute_radon_script(item.clone(), subscript.as_slice(), context, settings)?;
+
+        results.push(item.clone());
+        reports.push(report);
     }
 
-    Ok(RadonArray::from(result).into())
+    if let Stage::Retrieval(metadata) = &mut context.stage {
+        let mut partial_results = vec![];
```

Comment:
I think this piece of code is duplicated in `ArrayMap` and `ArrayFilter`. We can extract it into a function and give it a name.

The benefit of using iter and collect instead of for loop and push is that you get a capacity estimate for free. So try to do some iterator magic, or simply use `Vec::with_capacity(subscript.len() + 1)` here.

It should be possible to implement this without using clone, maybe using `std::mem::take` or something like that.

I don't think the `if let` are needed, the subscripts results must all have the same length, right? And report.partial_results cannot be `None`?. In that case just use unwrap or expect, that will help us to find issues in tests.

Reply:
All of these comments were very well received and have been already addressed. Thanks bro.

The only thing I couldn't address is about using `clone`. I will always end up having a `&mut RadonTypes` from which it is impossible to `mem::take` because `RadonTypes` does not implement `Default` and, even if it did, constructing a default value just to replace the mutable reference by means of `mem::take` will roughly equal the cost of simply cloning it.

Reply:
Ok, good enough.

> constructing a default value just to replace the mutable reference by means of mem::take will roughly equal the cost of simply cloning it.

That's not true if the type is a large `RadonArray` or `RadonMap`.